pkg/network/usm/ebpf_gotls_helpers.go

```diff
@@ -8,15 +8,134 @@
 package usm
 
 import (
+	"debug/elf"
 	"errors"
 	"fmt"
+	"os"
 	"reflect"
+	"time"
 	"unsafe"
 
+	"github.com/cilium/ebpf"
+	"golang.org/x/sys/unix"
+
+	"github.com/DataDog/datadog-agent/pkg/ebpf/uprobes"
 	"github.com/DataDog/datadog-agent/pkg/network/go/bininspect"
 	"github.com/DataDog/datadog-agent/pkg/network/protocols/http/gotls"
+	libtelemetry "github.com/DataDog/datadog-agent/pkg/network/protocols/telemetry"
+	"github.com/DataDog/datadog-agent/pkg/network/usm/utils"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
+// GoTLSBinaryInspector is a BinaryInspector that inspects Go binaries, dealing with the specifics of Go binaries
+// such as the argument passing convention and the lack of uprobes
+type GoTLSBinaryInspector struct {
+	structFieldsLookupFunctions map[bininspect.FieldIdentifier]bininspect.StructLookupFunction
+	paramLookupFunctions        map[string]bininspect.ParameterLookupFunction
+
+	// eBPF map holding the result of binary analysis, indexed by binaries'
+	// inodes.
+	offsetsDataMap *ebpf.Map
+
+	// binAnalysisMetric handles telemetry on the time spent doing binary
+	// analysis
+	binAnalysisMetric *libtelemetry.Counter
+
+	// binNoSymbolsMetric counts Golang binaries without symbols.
+	binNoSymbolsMetric *libtelemetry.Counter
+}
+
+// Ensure GoTLSBinaryInspector implements BinaryInspector
+var _ uprobes.BinaryInspector = &GoTLSBinaryInspector{}
+
+// Inspect extracts the metadata required to attach to a Go binary from the ELF file at the given path.
+func (p *GoTLSBinaryInspector) Inspect(fpath utils.FilePath, requests []uprobes.SymbolRequest) (map[string]bininspect.FunctionMetadata, bool, error) {
```

Comment:
why do we need both a boolean and error return arguments?
seems like mutually exclusive

Reply:
The idea was to distinguish between the case where a binary is incompatible and can't be probed (e.g., a x86 binary in ARM) and actual errors reading/parsing the file. Although to be honest I'm not sure how useful is that distinction now, as in the end the only thing the `UprobeAttacher` does is log both errors, just with a different message.

Reply:
you can create (and return) a specific error for that use case

Reply:
Removed in a separate PR: https://github.com/DataDog/datadog-agent/pull/29990